lightning/src/routing/scoring.rs

```diff
@@ -460,25 +458,466 @@ impl<T: Time> Readable for ChannelFailure<T> {
+}
+
+/// The likelihood of an event occurring.
+enum Probability {
+	Zero,
+	One,
+	Ratio { numerator: u64, denominator: u64 },
+}
+
+impl<G: Deref<Target = NetworkGraph>, T: Time> ProbabilisticScorerUsingTime<G, T> {
+	/// Creates a new scorer using the given scoring parameters for sending payments from a node
+	/// through a network graph.
+	pub fn new(
+		params: ProbabilisticScoringParameters, node_pubkey: &PublicKey, network_graph: G
+	) -> Self {
+		Self {
+			params,
+			node_id: NodeId::from_pubkey(node_pubkey),
+			network_graph,
+			channel_liquidities: HashMap::new(),
+		}
+	}
+
+	#[cfg(test)]
+	fn with_channel(mut self, short_channel_id: u64, liquidity: ChannelLiquidity<T>) -> Self {
+		assert!(self.channel_liquidities.insert(short_channel_id, liquidity).is_none());
+		self
+	}
+}
+
+impl Default for ProbabilisticScoringParameters {
+	fn default() -> Self {
+		Self {
+			liquidity_penalty_multiplier_msat: 1000,
+			liquidity_offset_half_life: Duration::from_secs(3600),
+		}
+	}
+}
+
+impl<T: Time> ChannelLiquidity<T> {
+	#[inline]
+	fn new() -> Self {
+		Self {
+			min_liquidity_offset_msat: 0,
+			max_liquidity_offset_msat: 0,
+			last_updated: T::now(),
+		}
+	}
+
+	/// Returns a view of the channel liquidity directed from `source` to `target` assuming
+	/// `capacity_msat`.
+	fn as_directed(
+		&self, source: &NodeId, target: &NodeId, capacity_msat: u64, half_life: Duration
+	) -> DirectedChannelLiquidity<&u64, T, &T> {
+		let (min_liquidity_offset_msat, max_liquidity_offset_msat) = if source < target {
+			(&self.min_liquidity_offset_msat, &self.max_liquidity_offset_msat)
+		} else {
+			(&self.max_liquidity_offset_msat, &self.min_liquidity_offset_msat)
+		};
+
+		DirectedChannelLiquidity {
+			min_liquidity_offset_msat,
+			max_liquidity_offset_msat,
+			capacity_msat,
+			last_updated: &self.last_updated,
+			now: T::now(),
+			half_life,
+		}
+	}
+
+	/// Returns a mutable view of the channel liquidity directed from `source` to `target` assuming
+	/// `capacity_msat`.
+	fn as_directed_mut(
+		&mut self, source: &NodeId, target: &NodeId, capacity_msat: u64, half_life: Duration
+	) -> DirectedChannelLiquidity<&mut u64, T, &mut T> {
+		let (min_liquidity_offset_msat, max_liquidity_offset_msat) = if source < target {
+			(&mut self.min_liquidity_offset_msat, &mut self.max_liquidity_offset_msat)
+		} else {
+			(&mut self.max_liquidity_offset_msat, &mut self.min_liquidity_offset_msat)
+		};
+
+		DirectedChannelLiquidity {
+			min_liquidity_offset_msat,
+			max_liquidity_offset_msat,
+			capacity_msat,
+			last_updated: &mut self.last_updated,
+			now: T::now(),
+			half_life,
+		}
+	}
+}
+
+impl<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> DirectedChannelLiquidity<L, T, U> {
+	/// Returns the success probability of routing the given HTLC `amount_msat` through the channel
+	/// in this direction.
+	fn success_probability(&self, amount_msat: u64) -> Probability {
+		let max_liquidity_msat = self.max_liquidity_msat();
```

Comment:
regarding one of my earlier comments I tried to chase down the logic of how this value is being derived and as far as I can tell this is the channel capacity or a potentially decayed learned bound but it is not constrained by `htlc_max_msat` like in the effective capacity data structure. I just leave the comment as I would like to suggest a second pair of eyes double checks this.

Reply:
I believe this value comes from the `as_directed` call in `channel_penalty_msat` which gets its information from the `available_liquidity` in the router, which calls `DirectedChannelInfo::effective_capacity` which is bound by the `htlc_maximum_msat`. Not quite sure whether you were saying this *should* or *should not* be, but it is.

Reply:
yeah I addressed this in another comment. I say for probability computation you should use the actual capacity (reduced by potential knowlege about existing liquidity) instead of capping this at `htlc_maximum_msat`. Using `htlc_maxiumum_msat` will result in potentially strange result (see this comment: https://github.com/lightningdevkit/rust-lightning/pull/1227#discussion_r788000465)